### PR TITLE
Add privacy/security notes to specification.

### PR DESCRIPTION
**Pull request by msporny**
This PR adds a few privacy/security notes to the specification as issues... the issues will need to be refined into specification text.

***
<a href="https://pr-preview.s3.amazonaws.com/msporny/rdf-canon/pull/77.html" title="Last updated on Feb 4, 2023, 9:52 PM UTC (4268cb2)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/rdf-canon/77/5e55147...msporny:4268cb2.html" title="Last updated on Feb 4, 2023, 9:52 PM UTC (4268cb2)">Diff</a>

### CODE REVIEW

**Comment by gkellogg** [comment]
The Echidna failure with undefined terms is unrelated to this PR and likely due to some change in infrastructure which requires further investigation.

**Review by gkellogg — CHANGES_REQUESTED** [review]
Can we reference specific issues on the repo using `data-number`? If we don't have existing issues, they should be created anyway.

Otherwise, looks good to me.

File: spec/index.html [hunk 1]
```diff
@@ -2629,13 +2629,57 @@ <h3>Algorithm</h3>
...
 <section id="security-considerations">
   <h2>Security Considerations</h2>
 
-  <p class="ednote">TBD</p>
+  <section>
+    <h3>Dataset Poisoning</h3>
+
+    <p class="issue" title="Attackers can construct poison Datasets">
+Add text that warns that attackers can construct Datasets which are known to
+take large amounts of compute time. The algorithm has a mechanism to detect
```
Comment (dlongley) [inline spec/index.html:2656]:
```suggestion
take large amounts of compute time to canonize. The algorithm has a mechanism to detect
```

File: spec/index.html [hunk 2]
```diff
@@ -2629,13 +2629,57 @@ <h3>Algorithm</h3>
...
+and prevent this sort of abuse, but implementers need to ensure that they
+think holistically about their system such as what happens if they don't have
+rate limiting on a service exposed to the Internet and they are the subject of
+a DDoS. Default mechanisms that prevent excessive use of compute when an
+attacker sends a poisoned Dataset might be different from system to system.
```
Comment (dlongley) [inline spec/index.html:2661]:
```suggestion
attacker sends a poisoned dataset might be different from system to system.
```

File: spec/index.html [hunk 3]
```diff
@@ -2629,13 +2629,57 @@ <h3>Algorithm</h3>
...
+Add text that warns implementers that, while the algorithm has a mathematical
+proof associated with it that has had peer review, and while a W3C WG
+has reviewed the algorithms and that there are multiple interoperable
+implementations, that a formal proof using a system such as Coq isn't available
+at this time. We are highly confident of the correctness of the algorithm,
+but we will not be able to say with 100% certainty that it is correct until
+we have a formal, machine-based verification of the proof. Any system that
+utilizes this canonicalization mechanism should have a backup canonicalization
+mechanism, such as JCS, ready in the event that an unrecoverable flaw is found
+in this algorithm.
```
Comment (dlongley) [inline spec/index.html:2679]:
```suggestion
utilizes this canonicalization mechanism should have a backup canonicalization
mechanism, such as JCS, or employ other mitigations, such as schema-based
validation, ready in the event that an unrecoverable flaw is found in this algorithm.
```

**Review by dlongley — APPROVED** [review]
Thanks! Approving w/suggestions.

File: spec/index.html [hunk 4]
```diff
@@ -2629,13 +2629,57 @@ <h3>Algorithm</h3>
 <section id="privacy-considerations">
   <h2>Privacy Considerations</h2>
 
-  <p class="ednote">TBD</p>
+  <section>
+    <h3>Selective Disclosure Schemes</h3>
+
+    <p class="issue" title="Dataset structure might reveal information">
```
Comment (gkellogg) [inline spec/index.html:2635]:
```suggestion
    <p class="issue" data-number="70" title="Dataset structure might reveal information">
```

File: spec/index.html [hunk 5]
```diff
@@ -2629,13 +2629,57 @@ <h3>Algorithm</h3>
...
 </section>
 
 <section id="security-considerations">
   <h2>Security Considerations</h2>
 
-  <p class="ednote">TBD</p>
+  <section>
+    <h3>Dataset Poisoning</h3>
+
+    <p class="issue" title="Attackers can construct poison Datasets">
```
[the remark this answers is not in the document]
Comment (gkellogg) [reply spec/index.html:2654]: Adding `data-number:

```suggestion
    <p class="issue" data-number="70" title="Attackers can construct poison datasets">
```

File: spec/index.html [hunk 6]
```diff
@@ -2629,13 +2629,57 @@ <h3>Algorithm</h3>
...
+a DDoS. Default mechanisms that prevent excessive use of compute when an
+attacker sends a poisoned Dataset might be different from system to system.
+    </p>
+  </section>
+
+
+  <section>
+    <h3>Formal Verification Incomplete</h3>
+
+    <p class="issue" title="Formal verification of algorithm is incomplete">
```
Comment (gkellogg) [inline spec/index.html:2669]:
```suggestion
    <p class="issue" data-number="70" title="Formal verification of algorithm is incomplete">
```

**Review by gkellogg — CHANGES_REQUESTED** [review]
Along with @dlongley's suggestions, I added `data-number="70"` to reference the driving issue.

File: spec/index.html [hunk 7]
```diff
@@ -2629,13 +2629,57 @@ <h3>Algorithm</h3>
...
 </section>
 
 <section id="security-considerations">
   <h2>Security Considerations</h2>
 
-  <p class="ednote">TBD</p>
+  <section>
+    <h3>Dataset Poisoning</h3>
+
+    <p class="issue" title="Attackers can construct poison datasets">
```
Comment (msporny) [inline spec/index.html:2654]:
```suggestion
    <p class="issue" data-number="70" title="Attackers can construct poison datasets">
```

**Review by gkellogg — APPROVED** [review]

**Review by yamdan — APPROVED** [review]
Thank you @msporny, I will give it a shot on adding something in the section of Selective Disclosure Schemes later; I think not only graph structure but also partial information of the literals could been leaked via canonical labels (or orders), as @dlongley already mentioned in #4 and #8.

File: spec/index.html [hunk 8]
```diff
@@ -2675,8 +2675,8 @@ <h3>Formal Verification Incomplete</h3>
 but we will not be able to say with 100% certainty that it is correct until
 we have a formal, machine-based verification of the proof. Any system that
 utilizes this canonicalization mechanism should have a backup canonicalization
-mechanism, such as JCS, ready in the event that an unrecoverable flaw is found
-in this algorithm.
+mechanism, such as JCS, or employ other mitigations, such as schema-based
```
Comment (TallTed) [inline spec/index.html:2678]:
```suggestion
mechanism, such as JCS, or other mitigations to employ, such as schema-based
```

**Review by peacekeeper — APPROVED** [review]
Thanks for moving this forward!

**Comment by gkellogg** [comment]
@msporny ready to merge?

**Comment by msporny** [comment]
> @msporny ready to merge?

Yep, I think we got everyone's change requests in. Please merge at your leisure. :)

**Pull request merged**